### PR TITLE
Add MaskedTensor passthrough: unfold, F.Unfold, F.Fold, stack

### PR DESCRIPTION
**Pull request by nowtryz**
Hi,
I noticed the `unfold` operator was missing on MaskedTensor.

I tested that my change works when calling unfold and backward on a `MaskedTensor` but I didn't find the tests for the dispatch of such operation. Where is it?

### CODE REVIEW

**Comment by albanD** [comment]
Most likely the right fix indeed.
cc @cpuhrsch what is the right way to test these PRs?

**Comment by cpuhrsch** [comment]
Hi @nowtryz - Thanks for sending the PR! The tests are in https://github.com/pytorch/pytorch/blob/main/test/test_maskedtensor.py .

**Comment by nowtryz** [comment]
Hi,

Is it possible to have a bit of help to understand how the test framework
works? Or a link to a documentation?
When I used the debugger, I did not see the pass through dispatch being
called.

Thank you

**Comment by cpuhrsch** [comment]
@nowtryz - You should be able to run this with `pytest test/test_maskedtensor.py`. You'll need to extend the tests to also cover fold/unfold. For example see [`test_prod`](https://github.com/pytorch/pytorch/blob/e3627d05e789c6cd20eeccce89dd733f02a808a9/test/test_maskedtensor.py#L714-L725. Since fold/unfold is neither unary, binary or a reduction I'd add a new test under `TestBasics` (not that it's a basic op, but because it doesn't fit the other buckets).

**Comment by cpuhrsch** [comment]
@nowtryz - Yes, exactly. A `test_fold` method for example.

**Comment by nowtryz** [comment]
Hi,

From what I understand, `torch.nn.Fold` has a specific semantic that requires a custom masked implementation: what should happen when I fold a combination of specified and unspecified values? Should I sum only the specified values or consider the fold unspecified? Consequently, I feel that this is out of the perimeter of this pull request. Feel free to correct me if I'm wrong.

In the meantime, I added support for `torch.nn.Unfold`/`torch.nn.functional.unfold`.

**Comment by cpuhrsch** [comment]
@nowtryz - Hm, on reduction semantics see details here: https://pytorch.org/tutorials/prototype/maskedtensor_advanced_semantics#reduction-semantics and also https://github.com/pytorch/rfcs/pull/27 .

One way to derive semantics of composites such as Fold (or matmul) is by inferring them from the semantics of its components.

Or said different (and maybe less confusingly), if you were to implement Fold in terms of existing MaskedTensor operations (if that's possible), what would it do?

These kind of discussions are really important (in my very biased opinion) to separate out all the subtly different use cases: raggedness (`tensor[i].size()` and `tensor[j].size()` may not match for `i != j`), sparsity (storage layouts for Tensors with lots of zeros) and masking (logical elements that have a position, but no value).

So, for example, if you mask out an element it doesn't have a value, but it does still have a position. Should that influence the denominator in `variance`? If you do sparsity, well, `0` is still a value. So if you do softmax, you want to include that (but often people actually would rather it mean `-inf` because really they want to mask it out). But then if you do raggedness, well you just want to actually describe an entirely different shape. You want to entirely remove elements (e.g. sentences of different lengths) and not have their individual shape influence operations that are truly batched (e.g. parallel or commute with vmap along dimension 0).

**Comment by nowtryz** [comment]
Hi @cpuhrsch, 

I'm trying to catch with the Fold/Unfold operators as they are different from Tensor.unfold. Looking at the operator's documentation, I understand the masked semantics of this operator like this:

If we have blocks of size $L$, then the folded cell would have the value $L \times x$, $x$ being the initial value of an equivalent `Unfold` operation yielding the given block. In that sens, if we have $N$ specified elements, $N \le L$, my understanding is that the reduction of the given block should be

$$\frac{L}{N} \sum_{n=1}^{N} x_n $$

If I am correct, then I am not sure how to implement if efficiently. My first idea (to reuse the existing ATen operator without further modifications) is the following:
1. the new mask is computed by passing the original one through the operator, which yields `True` for each block where at least one value was specified;
2. we compute the normal folded version of the data after filling unspecified values with 0s, let's call it "folded data";
3. the original mask is converted to an integer type and passed through the operator, which yields the $N$ values mentioned earlier, let's call it "divisors";
4. the result would be `res = input.size(-1) * divisors * folded_data`.

I don't know if it is more optimized, but I can also combine (1) and (3) by performing (3) and converting the result back to boolean to obtain the new mask.

What do you think?

**Comment by nowtryz** [comment]
Hi @cpuhrsch @mruberry,

Could we merge this PR and I will open a new one for the Fold when I have more time?

Many thanks

**Comment by cpuhrsch** [comment]
@nowtryz I had to unexpectedly go on leave due to family reasons, but let's revisit this. Are there tests for im2col or col2im that you could add the additional dtype to? Otherwise this looks fine, thank you for sending this :)

**Comment by nowtryz** [comment]
Hi @cpuhrsch,

No problem!  If I understand correctly, updating torch/testing/_internal/common_methods_invocations.py should have generated the correct tests for im2col and col2im. Though the tests fail for these functions on cuda with the dtype I added. I have to check the cuda kernel but the fix should be easy (however I do not know cuda)

**Comment by cpuhrsch** [comment]
@nowtryz - Thanks for double checking, please ping again if you need more pointers.

**Comment by nowtryz** [comment]
Hi @cpuhrsch,

I added the support for bool on cuda devices but I don't find the tests for im2col nor col2im. Tests have been generated for `nn_functional_unfold_cuda_bool` which uses im2col and col2im but I am not sure if it is sufficient. Maybe you or one of the code owners could guide me to find these tests?

Many thanks

**Comment by cpuhrsch** [comment]
Hello @nowtryz - Hm, what you did is likely to be sufficient. However, I can't seem to find where the test uses col2im. Do you have a pointer to that? Thank you for the PR(s)! :)

**Comment by nowtryz** [comment]
Hi @cpuhrsch,

To my understanding, as part of the tests using `torch/testing/_internal/common_methods_invocations.py`'s `OpInfo`s, there are parametrized tests testing the ops with all supported dtypes as well as supposedly unsupported dtypes. This test verifies that the function fails when called with an unsupported dtype and works when called with tensors generated for the supported dtypes.

I had to compile to compile locally with GPU support, was pretty long. But you can run the following:
```sh
pytest test/test_ops.py -vk "unfold and bool"
```
which produces boolean tests for nn unfold. I previously saw that those tests also included a backward pass somewhere.

**Comment by cpuhrsch** [comment]
Ok, great, thank you for verifying @nowtryz.

**Review by cpuhrsch — DISMISSED** [review]

**Comment by nowtryz** [comment]
@ZainRizvi the test fail is caused by #128637, not this PR ([as seen here](https://github.com/pytorch/pytorch/actions/runs/10604690693/job/29393478235#step:20:3075)

**Comment by nowtryz** [comment]
Hi @cpuhrsch!
Last failure didn't seem to be related to the code of this PR and passed locally.

**Review by cpuhrsch — APPROVED** [review]

**Comment by clee2000** [comment]
I believe this caused a memory leak in test_maskedtensor.py::TestBasicsCUDA::test_stack_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/10802287000/job/29965265434) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a15aabc975ccab5d717dd0702c12ee75a9ff58c3), @nowtryz could you take a look?

**Comment by cpuhrsch** [comment]
@clee2000 - I think we'll need to revert this again then.

**Comment by nowtryz** [comment]
@clee2000 That's strange, the test is just doing a passthrough to the stack function, there is no change in the cuda kernels as part of this test, I will need to take a deeper look
